src/SelfGravity/CPU_Poisson/CPU_ComputeGREP.cpp

```diff
@@ -0,0 +1,203 @@
+                                     1.0 + ( 0.25 * SQR( VrAve->Data[i] + VrAve->Data[i+1] )
+                                           - 2.0  * NEWTON_G * Mass_TOV[i]      / EdgeL[i+1]  ) / c2 ) );
+
+      Gamma_TOV[NBin-1] = SQRT( MAX( TINY_NUMBER,
+                                     1.0 + (        SQR( VrAve->Data[NBin-1] )
+                                           - 2.0 *  NEWTON_G * Mass_TOV[NBin-1] / MaxRadius   ) / c2 ) );
+
+
+//    check whether the tolerance is satisfied
+      bool Pass = true;
+      for (int i=0; i<NBin; i++)
+      {
+         double error = FABS( Mass_TOV_USG[i] - Mass_TOV[i] ) / Mass_TOV[i];
+
+         if ( error > tolerance )
+         {
+            Pass = false;
+            break;
+         }
+      }
+
+      if ( Pass )   break;
+      else          for (int i=0; i<NBin; i++)   Mass_TOV_USG[i] = Mass_TOV[i];
+
+
+//    additional information for debug if no convegent solution is found
+      if ( !NIter  &&  !Pass )
+      {
+         if ( MPI_Rank == 0 )
+         {
+            printf("\n============================================================\n");
+            printf("GREP_CENTER_METHOD : %d\n",                   GREP_CENTER_METHOD);
+            printf("Center             : %13.7e %13.7e %13.7e\n", DensAve->Center[0], DensAve->Center[1], DensAve->Center[2]);
+            printf("MaxRadius          : %13.7e\n",               DensAve->MaxRadius);
+            printf("LogBin             : %d\n",                   DensAve->LogBin);
+            printf("LogBinRatio        : %13.7e\n",               DensAve->LogBinRatio);
+            printf("Num of Iterations  : %d\n",                   GREP_MAXITER - NIter);
+            printf("NBin               : %d\n",                   NBin);
+            printf("============================================================\n");
+            printf("%5s %9s %10s %10s %10s %10s %10s %10s %10s %10s %10s %10s\n",
+                   "Bin", "NCell", "Radius", "Dens", "Engy", "Vr", "Pressure",
+                   "Mass_NW", "Mass_TOV", "Mass_TOV_old", "Error_rel", "Gamma_TOV");
+
+            for (int i=0; i<NBin; i++)
+               printf("%5d %9ld %10.2e %10.2e %10.2e %10.2e %10.2e %10.2e %10.2e %12.2e %10.2e %10.2e\n",
+                      i, DensAve->NCell[i], Radius[i],
+                      DensAve->Data[i], EngyAve->Data[i], VrAve->Data[i], PresAve->Data[i],
+                      Mass_NW[i], Mass_TOV[i], Mass_TOV_USG[i],
+                      FABS( Mass_TOV_USG[i] - Mass_TOV[i] ) / Mass_TOV[i], Gamma_TOV[i]);
+
+            Aux_Error( ERROR_INFO, "Too many iterations in constructing effective potential\n" );
+         }
+      }
+
+   } // while ( NIter-- )
+
+
+// copy Profile parameters to Phi_eff
+   Phi_eff->NBin        = DensAve->NBin;
+   Phi_eff->LogBin      = DensAve->LogBin;
+   Phi_eff->LogBinRatio = DensAve->LogBinRatio;
+   Phi_eff->MaxRadius   = DensAve->MaxRadius;
+   Phi_eff->AllocateMemory();
```

Comment:
- Yes, Phi_eff is free's and allocated again when calling `AllocateMemory()` since the bin could change with time.
- It is feasible, and I think we also need to allocate the others (e.g., `DensAve`) in `Init_GREP()` and fill the empty bins with interpolated values. However, this will also increase the time for constructing GREP, and I am not sure if it will save time.
- I add a new function `End_MemFree_GREP()` to free all the Profile_t objects invoking in `End_GREP_MigrationTest()`.